src/stream/src/from_proto/hash_join.rs

```diff
@@ -197,43 +199,57 @@ impl<S: StateStore> HashKeyDispatcher for HashJoinExecutorDispatcherArgs<S> {
-                    self.lru_manager,
-                    self.is_append_only,
-                    self.metrics,
-                    self.chunk_size,
-                    self.high_join_amplification_threshold,
+            ($join_type:ident, $join_encoding:ident) => {
+                Ok(
+                    HashJoinExecutor::<K, S, { JoinType::$join_type }, $join_encoding>::new(
+                        self.ctx,
+                        self.info,
+                        self.source_l,
+                        self.source_r,
+                        self.params_l,
+                        self.params_r,
+                        self.null_safe,
+                        self.output_indices,
+                        self.cond,
+                        self.inequality_pairs,
+                        self.state_table_l,
+                        self.degree_state_table_l,
+                        self.state_table_r,
+                        self.degree_state_table_r,
+                        self.lru_manager,
+                        self.is_append_only,
+                        self.metrics,
+                        self.chunk_size,
+                        self.high_join_amplification_threshold,
+                    )
+                    .boxed(),
                 )
-                .boxed())
             };
         }
-        match self.join_type_proto {
-            JoinTypeProto::AsofInner
-            | JoinTypeProto::AsofLeftOuter
-            | JoinTypeProto::Unspecified => unreachable!(),
-            JoinTypeProto::Inner => build!(Inner),
-            JoinTypeProto::LeftOuter => build!(LeftOuter),
-            JoinTypeProto::RightOuter => build!(RightOuter),
-            JoinTypeProto::FullOuter => build!(FullOuter),
-            JoinTypeProto::LeftSemi => build!(LeftSemi),
-            JoinTypeProto::LeftAnti => build!(LeftAnti),
-            JoinTypeProto::RightSemi => build!(RightSemi),
-            JoinTypeProto::RightAnti => build!(RightAnti),
+
+        macro_rules! build_match {
+            ($(($join_type:ident)),*) => {
+                match (self.join_type_proto, self.join_encoding_type) {
+                    (JoinTypeProto::AsofInner, _)
+                    | (JoinTypeProto::AsofLeftOuter, _)
+                    | (JoinTypeProto::Unspecified, _)
+                    | (_, JoinEncodingTypeProto::Unspecified ) => unreachable!(),
+                    $(
+                        (JoinTypeProto::$join_type, JoinEncodingTypeProto::MemoryOptimized) => build!($join_type, MemoryEncoding),
+                        (JoinTypeProto::$join_type, JoinEncodingTypeProto::CpuOptimized) => build!($join_type, CpuEncoding),
+                    )*
+                }
+            };
+        }
+        build_match! {
+            (Inner),
+            (LeftOuter),
+            (RightOuter),
+            (FullOuter),
+            (LeftSemi),
+            (LeftAnti),
+            (RightSemi),
+            (RightAnti)
```

Comment:
Nit: Why do we need parenthesis here. Seems like this can be removed from the macro instead of `(Inner)`, can just be `Inner`.